dev-tools/omdb/src/bin/omdb/mgs/sensors.rs

```diff
@@ -0,0 +1,954 @@
+    let by_kind = if let Some(types) = &args.types {
+        let mut h = HashSet::new();
+
+        for t in types {
+            h.insert(match Sensor::from_string("", &t) {
+                None => bail!("invalid sensor kind {t}"),
+                Some(s) => s.kind,
+            });
+        }
+
+        Some(h)
+    } else {
+        None
+    };
+
+    let by_name = if let Some(named) = &args.named {
+        Some(named.into_iter().collect::<HashSet<_>>())
+    } else {
+        None
+    };
+
+    let info = match input {
+        SensorInput::MgsClient(ref mgs_client) => {
+            sp_info_mgs(mgs_client, args).await?
+        }
+        SensorInput::CsvReader(reader, position) => {
+            sp_info_csv(reader, position, args)?
+        }
+    };
+
+    let mut sensors_by_sensor = MultiMap::new();
+    let mut sensors_by_sensor_and_sp = HashMap::new();
+    let mut sensors_by_id = HashMap::new();
+    let mut sensors_by_sp = MultiMap::new();
+    let mut values = HashMap::new();
+    let mut work_by_sp = HashMap::new();
+
+    let mut current = 0;
+    let time = info.time;
+
+    for (sp_id, info) in info.info {
+        let mut sp_work = vec![];
+
+        for (device, sensors) in info.devices {
+            let mut device_work = vec![];
+
+            for (sensor, value) in sensors {
+                if let Some(ref by_kind) = by_kind {
+                    if by_kind.get(&sensor.kind).is_none() {
+                        continue;
+                    }
+                }
+
+                if let Some(ref by_name) = by_name {
+                    if by_name.get(&sensor.name).is_none() {
+                        continue;
+                    }
+                }
+
+                let id = SensorId(current);
+                current += 1;
+
+                sensors_by_id
+                    .insert(id, (sp_id, sensor.clone(), device.clone()));
+
+                if value.is_none() && args.verbose {
+                    eprintln!(
+                        "mgs: error for {sp_id:?} on {sensor:?} ({device:?})"
+                    );
+                }
+
+                sensors_by_sensor.insert(sensor.clone(), id);
+
+                let by_sp = sensors_by_sensor_and_sp
+                    .entry(sensor)
+                    .or_insert_with(|| HashMap::new());
+                by_sp.insert(sp_id, id);
+                sensors_by_sp.insert(sp_id, id);
+                values.insert(id, value);
+
+                device_work.push(id);
+            }
+
+            sp_work.push((device, device_work));
+        }
+
+        work_by_sp.insert(sp_id, sp_work);
+    }
+
+    Ok((
+        SensorMetadata {
+            sensors_by_sensor,
+            sensors_by_sensor_and_sp,
+            sensors_by_id,
+            sensors_by_sp,
+            work_by_sp,
+            start_time: args.start,
+            end_time: match args.end {
+                Some(end) => Some(end),
+                None => match args.duration {
```

Comment:
I'd probably use `.map` here to avoid the nested match:

```suggestion
                None => args.duration.map(|duration| time + duration),
```